File: internal_testsuites/typescript/src/testsuite/add_service_with_application_protocol_test/add_service_with_application_protocol.test.ts
import {createEnclave} from "../../test_helpers/enclave_setup";
import log from "loglevel";
import {ContainerConfig, ContainerConfigBuilder, PortSpec} from "kurtosis-sdk";
import {Port} from "kurtosis-sdk/build/core/kurtosis_core_rpc_api_bindings/api_container_service_pb";
import TransportProtocol = Port.TransportProtocol;

const TEST_NAME = "add-service-with-port-spec"
const IS_PARTITIONING_ENABLED = false
const SERVICE_ID = "service-with-port-spec"

const PORT_WITH_APP_PROTOCOL    = "port1"
const PORT_WITHOUT_APP_PROTOCOL = "port2"
const DOCKER_IMAGE = "docker/getting-started:latest"


jest.setTimeout(180000)

test("Test add service with application protocol", TestAddServiceWithApplicationProtocol)

async function TestAddServiceWithApplicationProtocol () {
    const createEnclaveResult = await createEnclave(TEST_NAME, IS_PARTITIONING_ENABLED)
    if(createEnclaveResult.isErr()) { throw createEnclaveResult.error }
    const { enclaveContext, stopEnclaveFunction } = createEnclaveResult.value

    try {
        let usedPorts = new Map<string, PortSpec>()
        usedPorts.set( PORT_WITH_APP_PROTOCOL, new PortSpec(3333, TransportProtocol.UDP, "https"))
        usedPorts.set( PORT_WITHOUT_APP_PROTOCOL, new PortSpec(4444, TransportProtocol.TCP))

        let containerConfigs = new Map<string, ContainerConfig>()
        containerConfigs.set(SERVICE_ID, new ContainerConfigBuilder(DOCKER_IMAGE).withUsedPorts(usedPorts).build())

        const runResult = await enclaveContext.addServices(containerConfigs);
        if (runResult.isErr()) {
            log.error(`Unexpected error while adding serivce with id ${SERVICE_ID}`);
            throw runResult.error;
        }

        const results = runResult.value
        const serviceContext = results[0].get(SERVICE_ID)

        if (typeof serviceContext === "undefined") {
            log.error(`Service with Id ${SERVICE_ID} not found`)
            const error = results[1].get(SERVICE_ID)
            throw error
        }

        const portSpec = serviceContext.getPrivatePorts();

        const portSpecWithAppProtocol = portSpec.get(PORT_WITH_APP_PROTOCOL)
        expect(portSpecWithAppProtocol).toBeDefined()
        if (portSpecWithAppProtocol !== undefined) {
            expect(portSpecWithAppProtocol.transportProtocol).toEqual(TransportProtocol.UDP)
            expect(portSpecWithAppProtocol.number).toEqual(3333)
            expect(portSpecWithAppProtocol.maybeApplicationProtocol).toEqual("https")
        }

        const portSpecWithoutAppProtocol = portSpec.get(PORT_WITHOUT_APP_PROTOCOL)
        expect(portSpecWithoutAppProtocol).toBeDefined()
        if (portSpecWithoutAppProtocol !== undefined) {
            expect(portSpecWithoutAppProtocol.transportProtocol).toEqual(TransportProtocol.TCP)
            expect(portSpecWithoutAppProtocol.number).toEqual(4444)
            expect(portSpecWithoutAppProtocol.maybeApplicationProtocol).toEqual("")
        }
    } finally {
        stopEnclaveFunction()
    }

    jest.clearAllTimers()
}

File: internal_testsuites/golang/testsuite/add_service_with_application_protocol_test/add_service_with_application_protocol_test.go
package add_service_with_application_protocol_test

import (
	"context"
	"github.com/kurtosis-tech/kurtosis-cli/golang_internal_testsuite/test_helpers"
	"github.com/kurtosis-tech/kurtosis/api/golang/core/lib/services"
	"github.com/stretchr/testify/require"
	"testing"
)

const (
	testName               = "add-service-with-port-spec"
	isPartitioningEnabled  = false
	serviceIdStr           = "service-with-port-spec"
	portWithAppProtocol    = "port1"
	portWithOutAppProtocol = "port2"
	starlarkScript         = `
def run(plan):
	plan.add_service(
		service_id = "service-with-port-spec",
		config = struct(
			image = "docker/getting-started:latest",
			ports = {
				"port1": PortSpec(
					number = 4444,
					transport_protocol = "UDP",
					application_protocol = "http"
				),
				"port2": PortSpec(
					number = 3333,
					transport_protocol = "TCP"
				)
			},
		)
	)
`
)

func TestAddServiceWithApplicationProtocol(t *testing.T) {
	ctx := context.Background()
	serviceId := services.ServiceID(serviceIdStr)

	// ------------------------------------- ENGINE SETUP ----------------------------------------------
	enclaveCtx, _, destroyEnclaveFunc, err := test_helpers.CreateEnclave(t, ctx, testName, isPartitioningEnabled)
	defer func() {
		err := destroyEnclaveFunc()
		require.Nil(t, err)
	}()
	require.NoError(t, err, "An error occurred creating an enclave")

	// -------------------------------------- SCRIPT RUN -----------------------------------------------
	runResult, err := enclaveCtx.RunStarlarkScriptBlocking(ctx, starlarkScript, "", false)
	require.NoError(t, err, "An unexpected error occurred while running Starlark script")
	require.Empty(t, runResult.InterpretationError, "An unexpected error occurred while interpreting Starlark script")
	require.Empty(t, runResult.ValidationErrors, "An unexpected error occurred while validating Starlark script")
	require.Empty(t, runResult.ExecutionError, "An unexpected error occurred while executing Starlark script")

	// ------------------------------------ TEST ASSERTIONS ---------------------------------------------
	service, err := enclaveCtx.GetServiceContext(serviceId)
	require.NoError(t, err, "An unexpected error occurred while getting service")
	ports := service.GetPrivatePorts()
	portSpecWithAppProtocol := ports[portWithAppProtocol]
	portSpecWithoutAppProtocol := ports[portWithOutAppProtocol]

	require.Equal(t, uint16(4444), portSpecWithAppProtocol.GetNumber())
	require.Equal(t, "http", portSpecWithAppProtocol.GetMaybeApplicationProtocol())
	require.Equal(t, services.TransportProtocol_UDP, portSpecWithAppProtocol.GetTransportProtocol())

	require.Equal(t, uint16(3333), portSpecWithoutAppProtocol.GetNumber())
	require.Equal(t, "", portSpecWithoutAppProtocol.GetMaybeApplicationProtocol())
	require.Equal(t, services.TransportProtocol_TCP, portSpecWithoutAppProtocol.GetTransportProtocol())
}
